bin/picom-trans

```diff
@@ -46,194 +54,306 @@ print_usage() {
-opacity=$(echo "$opacity" | xargs | sed 's/%//g')
+  while test $# -gt 0; do
+    # 15 +15 -15 are opacity values
+    # filter them out here, otherwise getopts will complain
+    if echo "$1" | grep -qE '^[+-]?[[:digit:]]+%?$'; then
+      target_opacity=$1
+      shift
+      continue
+    fi
 
-# Validate opacity value
-if test -z "$action" && ! echo "$opacity" | grep -qE '^[+-]?[0-9]+$'; then
-  echo "Invalid opacity specified: $opacity."
-  exit 1
-fi
+    OPTIND=1
+    while getopts ':ho:gdtrscn:w:' OPTION
+    do
+      case "$OPTION" in
+        (h) print_usage; exit 0;;
+        (o) target_opacity="$OPTARG";;
+        (g) action=get;;
+        (d) action=delete;;
+        (t) action=toggle;;
+        (r) action=reset;;
+        (s) winidtype=;         winid=;;
+        (c) winidtype=-id;      winid="$(get_focused_window_id)";;
+        (n) winidtype=-name;    winid="$OPTARG";;
+        (w) winidtype=-id;      winid="$OPTARG";;
+        (\?)
+          # This might not be an invalid argument.
+          # It could be an opacity value.
+          #
+          # If it is an opacity value,  just break.
+          # It will be handled at next iteration of the loop.
+          if echo $OPTARG | grep -q '[[:digit:]]'
+          then break
+          else echo "$EXE_NAME: illegal option: $1" >&2 && exit 1
+          fi
+      esac
 
-# Reset opacity for all windows
-if test x"$action" = x'reset'; then
-  xwininfo -root -tree \
-  | sed -n 's/^     \(0x[[:xdigit:]]*\).*/\1/p' \
-  | while IFS=$(printf '\n') read wid; do
-    xprop -id "$wid" -remove _NET_WM_WINDOW_OPACITY
+      # If getopts goes to the second argument, stop
+      test $OPTIND -ge 3 && break
+    done
+    shift
   done
-  exit 0
-fi
+} #}
 
-# Get ID of the target window
-if test -z "$wprefix"; then
-  treeout=$(xwininfo -children -frame)
-else
-  test "$wprefix" = '-id' \
-    && ! echo "$window" | grep -Eiq '^[[:space:]]*(0x[[:xdigit:]]+|[[:digit:]]+)[[:space:]]*$' \
-    && echo 'Bad window ID.' && exit 1
-  treeout=$(xwininfo -children $wprefix "$window")
-fi
 
-wid=$(echo "$treeout" | sed -n 's/^xwininfo:.*: \(0x[[:xdigit:]]*\).*$/\1/p')
+# NOTE: The code inside the following function has been taken directly from the
+# old version of picom-trans.  Tread cautiously!
+function get_target_window_id
+{ #{
 
-if test -z "$wid"; then
-  echo 'Failed to find window.'
-  exit 1
-fi
+  # Get ID of the target window
+  if test -z "$winidtype"; then
+    treeout=$(xwininfo -children -frame)
+  else
+    test "$winidtype" = '-id' \
+      && ! echo "$winid" | grep -Eiq '^[[:space:]]*(0x[[:xdigit:]]+|[[:digit:]]+)[[:space:]]*$' \
+      && echo 'Bad window ID.' && exit 1
+    treeout=$(xwininfo -children $winidtype "$winid")
+  fi
 
-# Make sure it's not root window
-if echo "$treeout" | fgrep -q 'Parent window id: 0x0'; then
-  echo 'Cannot set opacity on root window.'
-  exit 1
-fi
+  wid=$(echo "$treeout" | sed -n 's/^xwininfo:.*: \(0x[[:xdigit:]]*\).*$/\1/p')
```

Comment:
We should be able to use `sed -nE 's/^xwininfo:.*: (0x[[:xdigit:]]*).*$/\1/p'` without breaking anything, if I am not mistaken. This allows us to use non-escaped braces `()` for the groups. Having said that, it works fine the way it is.
Also applies to line 190 and 248 below.

Reply:
I don't know about (potential) performance differences. Since we don't require anything specific to ERE, this regex should be fine as is. I was just wondering if there is a preference for one or the other.

Reply:
I think I prefer using BRE whenever possible.

I say _"I think I prefer"_ because I have never really thought of it before...

:shrug:

Reply:
I couldn't measure any meaningful difference in my limited tests (on my hardware). I don't have a strong preference for either one.